department-of-veterans-affairs/caseflow
# Add error handling for decision issues that have no decision dates

Resolves [jira ticket 2009](https://vajira.max.gov/browse/CASEFLOW-2009): Maintain existing decision issue visibility when decision issue has no decision date.

### Description
[Stephen H. reached out from Bat Team](https://dsva.slack.com/archives/CJL810329/p1626893293196400) because a user was not able to view decision issues that have no decision date. He asked if it was incorrect for a decision issue to lack a decision date, or if [the line that caused this bug](https://github.com/department-of-veterans-affairs/caseflow/blob/master/app/policies/appeal_decision_issues_policy.rb) needed to be changed.

I have read _a lot_ of documentation, but I can't determine if decision issues should always have decision dates.  I will reach out to a more senior Caseflow team member about this.  In the meantime, I added error handling and refactored the decision issues policy.  Now, if the `restrict_poa_visibility` feature toggle is enabled, a VSO POA will not be able to view decision issues that have no decision dates.

### Acceptance Criteria
- [ ] Automated tests pass
- [ ]  If user is a VSO POA and the `restrict_poa_visibility` feature toggle is on, user does not see a 500 when they attempt to view a decision issue that has no decision date
- [ ] If user is a VSO POA and the `restrict_poa_visibility` feature toggle is on, user will not see any decision issues that have no decision date
- [ ] If user is a VSO POA and the `restrict_poa_visibility` feature toggle is off, user will see any decision issues that have no decision date
- [ ] If user does not have a vso role, user will see any decision issues that have no decision date (This functionality existed before I made this ticket)

### Testing Plan
1. Log in as a user who both is a poa and has a vso role.
2. Create an appeal that has decision issues with a decision date in the past, a decision date in the future, and no decision date at all.
`FactoryBot.create(:appeal, :with_decision_issue, :decision_issue_with_future_date, :decision_issue_with_no_decision_date)`
3. Navigate to the appeal case details page and confirm that you can only see the decision issues with a decision date in the past.
<img width="1789" alt="VSOUserSeesDecisionIssues" src="https://user-images.githubusercontent.com/80347702/126672179-c2b0c04a-8d23-4681-a856-cf718f630fb6.png">
4. Switch to being a user who does not have a vso role.
5. Navigate to the appeal case details page and confirm that you can see all decision issues, regardless of decision date.
<img width="1789" alt="NonVSOUserSeesDecisionIssues" src="https://user-images.githubusercontent.com/80347702/126672332-3550dcfb-b1f4-484a-94ba-70273bb43914.png">
